### PR TITLE
remove chrono dep entirely

### PR DESCRIPTION
**Pull request by reaperhulk**

### CODE REVIEW

**Comment by alex** [comment]
a) I think we should have a common datetime struct that these both use.
b) Needs accessors so pyca/cryptography can get the fields

File: src/types.rs [hunk 1]
```diff
@@ -810,20 +808,88 @@ fn push_four_digits(dest: &mut WriteBuf, val: u16) -> WriteResult {
...
-        if v.year() >= 2050 || v.year() < 1950 {
-            return None;
+    pub fn new(
+        year: u16,
+        month: u8,
+        day: u8,
+        hour: u8,
+        minute: u8,
+        second: u8,
+    ) -> ParseResult<UtcTime> {
```
Comment (alex) [inline src/types.rs:881]: I think this should just take a datetime.

File: src/types.rs [hunk 2]
```diff
@@ -810,20 +808,88 @@ fn push_four_digits(dest: &mut WriteBuf, val: u16) -> WriteResult {
...
+        if year >= 2050 {
+            return Err(ParseError::new(ParseErrorKind::InvalidValue));
+        }
+        let dt = DateTime::new(year, month, day, hour, minute, second);
+        match dt {
```
Comment (alex) [inline src/types.rs:886]: This can just be `Ok(UtcTime(DateTime::new(...)?))`

File: src/types.rs [hunk 3]
```diff
@@ -889,19 +944,36 @@ impl SimpleAsn1Writable for UtcTime {
...
-        // Reject leap seconds, which aren't allowed by ASN.1. chrono encodes
-        // them as nanoseconds == 1000000.
-        if v.year() < 0 || v.nanosecond() >= 1_000_000 {
+    pub fn new(
+        year: u16,
+        month: u8,
+        day: u8,
+        hour: u8,
+        minute: u8,
+        second: u8,
```
Comment (alex) [inline src/types.rs:956]: Just take a datetime

File: src/types.rs [hunk 4]
```diff
@@ -810,20 +808,88 @@ fn push_four_digits(dest: &mut WriteBuf, val: u16) -> WriteResult {
...
+    ) -> ParseResult<DateTime> {
+        validate_date(year, month, day)?;
+        if minute > 59 || second > 59 {
```
Comment (alex) [inline src/types.rs:831]: Check hour as well?

File: src/types.rs [hunk 5]
```diff
@@ -889,19 +944,36 @@ impl SimpleAsn1Writable for UtcTime {
...
+        hour: u8,
+        minute: u8,
+        second: u8,
+    ) -> ParseResult<GeneralizedTime> {
+        if !(1..=12).contains(&month)
+            || !(1..=31).contains(&day)
+            || hour > 23
+            || minute > 59
+            || second > 59
+        {
```
Comment (alex) [inline src/types.rs:963]: Then this logic doesn't need to be here.

File: src/types.rs [hunk 6]
```diff
@@ -889,19 +944,36 @@ impl SimpleAsn1Writable for UtcTime {
...
             return Err(ParseError::new(ParseErrorKind::InvalidValue));
         }
-        Ok(GeneralizedTime(v))
+        Ok(GeneralizedTime(DateTime {
```
Comment (alex) [inline src/types.rs:966]: And this won't need to construct it.

File: src/types.rs [hunk 7]
```diff
@@ -810,20 +808,80 @@ fn push_four_digits(dest: &mut WriteBuf, val: u16) -> WriteResult {
...
+impl DateTime {
+    pub fn new(
+        year: u16,
+        month: u8,
+        day: u8,
+        hour: u8,
+        minute: u8,
+        second: u8,
+    ) -> ParseResult<DateTime> {
+        validate_date(year, month, day)?;
```
Comment (reaperhulk) [inline src/types.rs:832]: This is the only call site for `validate_date` now so we could choose to inline it in a follow-up PR.

File: src/types.rs [hunk 8]
```diff
@@ -810,20 +808,80 @@ fn push_four_digits(dest: &mut WriteBuf, val: u16) -> WriteResult {
     dest.push_byte(b'0' + (val % 10) as u8)
 }
 
+/// A structure representing a date and time. Wrapped by
```
Comment (alex) [inline src/types.rs:811]: Mention that they're always UTC.

File: src/types.rs [hunk 9]
```diff
@@ -810,20 +808,80 @@ fn push_four_digits(dest: &mut WriteBuf, val: u16) -> WriteResult {
...
 #[derive(Debug, Clone, PartialEq, Hash, Eq)]
-pub struct UtcTime(chrono::DateTime<chrono::Utc>);
+pub struct UtcTime(DateTime);
 
 impl UtcTime {
-    pub fn new(v: chrono::DateTime<chrono::Utc>) -> Option<UtcTime> {
-        if v.year() >= 2050 || v.year() < 1950 {
-            return None;
+    pub fn new(dt: DateTime) -> ParseResult<UtcTime> {
+        if dt.year() >= 2050 {
```
Comment (alex) [inline src/types.rs:878]: Where did the check for < 1950 go?

**Review by alex — APPROVED** [review]

**Pull request merged**